corehq/apps/cleanup/management/commands/fix_forms_with_missing_xmlns.py

```diff
@@ -0,0 +1,159 @@
+from couchforms.models import XFormInstance
+from dimagi.utils.couch.database import iter_docs
+from django.core.management.base import BaseCommand
+from optparse import make_option
+
+
+class Command(BaseCommand):
+    help = 'Fix forms with "undefined" xmlns'
+    args = '<log_path>'
+
+    option_list = BaseCommand.option_list + (
+        make_option(
+            '--dry-run',
+            action='store_true',
+            dest='dry_run',
+            default=False,
+            help="Don't do the actual modifications, but still log what would be affected"
+        ),
+    )
+
+    def handle(self, *args, **options):
+
+        dry_run = options.get("dry_run", True)
+        log_path = args[0].strip()
+
+        new_xmlnss = {}
+        num_fixed = 0
+        total, submissions = get_submissions_without_xmlns()
+
+        with open(log_path, "w") as f:
+            for i, xform_instance in enumerate(submissions):
+                if i % 500 == 0 and i != 0:
+                    print "Progress: {} of {} ({})".format(i, total, round(i / float(total), 2))
+                    print "We can fix {} of {} ({})".format(
+                        num_fixed, i, round(num_fixed / float(i), 2)
+                    )
+                if xform_instance.app_id in new_xmlnss:
+                    num_fixed += 1
+                    # We've already generated a new xmlns for this app
+                    set_xmlns_on_submission(xform_instance, new_xmlnss[xform_instance.app_id], dry_run)
+                    f.write("Set new xmlns on submission {}\n".format(xform_instance._id))
+                else:
+                    app = Application.get(xform_instance.app_id)
+                    forms_without_xmlns = get_forms_without_xmlns(app)
+                    if len(forms_without_xmlns) == 1:
+                        form = forms_without_xmlns[0]
+                        if not xforms_with_real_xmlns_possibly_exist(app._id, form):
+                            num_fixed += 1
+                            new_xmlns = generate_random_xmlns()
+                            new_xmlnss[xform_instance.app_id] = new_xmlns
+                            f.write("New xmlns for form {form.unique_id} in app {app._id} is {new_xmlns}\n".format(
+                                form=form,
+                                app=app,
+                                new_xmlns=new_xmlns
+                            ))
+                            set_xmlns_on_form(form, new_xmlns, app, dry_run)
+                            set_xmlns_on_submission(xform_instance, new_xmlnss[xform_instance.app_id], dry_run)
+                            f.write("Set new xmlns on submission {}\n".format(xform_instance._id))
+
+
+def get_submissions_without_xmlns():
+    # TODO: This view won't get archived forms etc
+    submissions = XFormInstance.get_db().view(
+        'couchforms/by_xmlns',
+        key="undefined",
+        include_docs=False,
+        reduce=False,
+    ).all()
```

Comment:
@emord I tried doing an ES query aggregated by app_id but es aggregations don't return the docs themselves, and the results are approximate